cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -38,14 +40,28 @@
 
   private final Set<String> startingPoints;
   private final Map<String, TransformDetail> transformDetailMap;
+  private final String sourceStageName;
 
   public TransformExecutor(Map<String, TransformDetail> transformDetailMap, Set<String> startingPoints) {
+    this(transformDetailMap, startingPoints, null);
+  }
+
+  public TransformExecutor(Map<String, TransformDetail> transformDetailMap, Set<String> startingPoints,
+                           @Nullable String sourceStageName) {
     this.transformDetailMap = transformDetailMap;
     this.startingPoints = startingPoints;
+    this.sourceStageName = sourceStageName;
   }
 
   public TransformResponse runOneIteration(IN input) throws Exception {
     for (String stageName : startingPoints) {
+      // If we have a non-null sourceStageName to match with stageName (in case of multiple inputs), then
+      // executeTransformation only if both are equal
+      if (!Strings.isNullOrEmpty(sourceStageName)) {
```

Comment:
this logic looks out of place.  It is cleaner if the TransformExecutor only executes transform and doesn't have to worry about multiple inputs. Instead of adding this sourceStageName, would it work if the caller just passes in a set containing just the sourceStageName as the thing in the startingPoints set?